### PR TITLE
Python updates

### PR DESCRIPTION
**Pull request by farizrahman4u**

### CODE REVIEW

**Comment by AlexDBlack** [comment]
Confirmed passing all datavec tests on linux

On Windows I'm getting this... 
https://gist.github.com/AlexDBlack/68fe556289fcd32a16b31304a3218a17

**Comment by AlexDBlack** [comment]
So, there's a legit issue here on Windows

If I delete javacpp cache and run: OK, passes
If I then run again: it fails as per the gist in my previous post

Somehow it's actually modifying the cache files?
here's `C:\Users\Alex\.javacpp\cache\numpy-1.17.3-1.5.2-windows-x86_64.jar\org\bytedeco` after the first run (which succeeds):
image

And here it is after the second run (without deleting the javacpp cache between runs; it fails):
image


Note the slightly different total file size - 16,550,850 vs. 16,550,691
Same number of files/folders though, so something is getting modified?

cc @saudet

**Comment by saudet** [comment]
Can you run a `diff -ru` on that and see which files are changed?

**Comment by AlexDBlack** [comment]
After running once, renaming cache to cache_singlerun, then running twice (on same default cache directory, with one pass and one failure as before):
```
diff --brief --recursive /mnt/c/Users/Alex/.javacpp/cache/numpy-1.17.3-1.5.2-windows-x86_64.jar/org/bytedeco/numpy /mnt/c/Users/Alex/.javacpp/cache_singlerun/numpy-1.17.3-1.5.2-windows-x86_64.jar/org/bytedeco/numpy
```
(note run via WSL 2 for diff tool)

I get the following:
```
Files /mnt/c/Users/Alex/.javacpp/cache/numpy-1.17.3-1.5.2-windows-x86_64.jar/org/bytedeco/numpy/windows-x86_64/python/numpy/core/__pycache__/overrides.cpython-37.pyc and /mnt/c/Users/Alex/.javacpp/cache_singlerun/numpy-1.17.3-1.5.2-windows-x86_64.jar/org/bytedeco/numpy/windows-x86_64/python/numpy/core/__pycache__/overrides.cpython-37.pyc differ
Files /mnt/c/Users/Alex/.javacpp/cache/numpy-1.17.3-1.5.2-windows-x86_64.jar/org/bytedeco/numpy/windows-x86_64/python/numpy/testing/_private/__pycache__/utils.cpython-37.pyc and /mnt/c/Users/Alex/.javacpp/cache_singlerun/numpy-1.17.3-1.5.2-windows-x86_64.jar/org/bytedeco/numpy/windows-x86_64/python/numpy/testing/_private/__pycache__/utils.cpython-37.pyc differ
```

**Comment by saudet** [comment]
Looks like it's Python corrupting its own cache here, nothing to do with JavaCPP. CPython isn't thread-safe by default, so it's probably us trying to use it from multiple threads without the proper locks: https://docs.python.org/3/c-api/init.html

**Comment by AlexDBlack** [comment]
There shouldn't be any multithreading here afaik, unless it's done internally in python executor... @farizrahman4u ?
Edit: also I didn't see any issues on linux
So maybe some sort of RC, between shutdown of one test (python executor) and startup of the next

**Comment by AlexDBlack** [comment]
Or something not properly closed between tests - @farizrahman4u ?
Unfortunately python isn't exactly my specialty, not sure how useful I'll be in debugging this at this point...

**Comment by farizrahman4u** [comment]
@AlexDBlack I dont think it was windows vs linux.. its different versions of numpy (you probably have different versions of numpy on windows and linux(wsl)). Should be fixed now, delete javacpp python cache and try again.

**Comment by saudet** [comment]
So this is a problem with NumPy? I'd recommend using the version distributed with the presets instead of depending on whatever versions are available on the system:
https://github.com/bytedeco/javacpp-presets/tree/master/numpy

**Comment by farizrahman4u** [comment]
@saudet no, we patch javacpp's numpy to support multi interpreter (limited support). The patch was written targeting an older version of numpy. Now it supports 1.7 too.

**Comment by saudet** [comment]
I'm not seeing numpy listed anywhere in the pom.xml files. How does that work?

File: datavec/datavec-python/src/main/java/org/datavec/python/PythonExecutioner.java [hunk 1]
```diff
@@ -496,32 +1036,109 @@ else if (obj instanceof String){
...
+    private static boolean _checkPatchApplied(String dest) {
+        try {
+            return FileUtils.readFileToString(new File(dest), "utf-8").startsWith("#patch");
+        } catch (IOException e) {
+            throw new RuntimeException("Error patching numpy");
+
+        }
+    }
+
+    private static void applyPatches() {
```
Comment (farizrahman4u) [inline datavec/datavec-python/src/main/java/org/datavec/python/PythonExecutioner.java:1128]: @saudet We patch numpy here.

File: datavec/datavec-python/src/main/java/org/datavec/python/PythonExecutioner.java [hunk 2]
```diff
@@ -496,32 +1036,109 @@ else if (obj instanceof String){
...
+        patches.add(new String[]{
+                "pythonexec/patch0.py",
+                evalString("__overrides_path")
+        });
+        patches.add(new String[]{
+                "pythonexec/patch1.py",
+                evalString("__random_path")
+        });
+
+        return patches;
```
Comment (farizrahman4u) [inline datavec/datavec-python/src/main/java/org/datavec/python/PythonExecutioner.java:1106]: @saudet There are the patches

File: datavec/datavec-python/src/main/resources/pythonexec/patch0.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+from numpy.core._multiarray_umath import (
+    add_docstring, implement_array_function, _get_implementing_args)
+from numpy.compat._inspect import getargspec
+
+
+ENABLE_ARRAY_FUNCTION = bool(
+    int(os.environ.get('NUMPY_EXPERIMENTAL_ARRAY_FUNCTION', 0)))
+
+
+ARRAY_FUNCTION_ENABLED = ENABLE_ARRAY_FUNCTION  # backward compat
```
Comment (farizrahman4u) [inline datavec/datavec-python/src/main/resources/pythonexec/patch0.py:17]: @saudet This var was renamed in numpy 1.7, so we alias it.

File: pom.xml [hunk 4]
```diff
@@ -294,6 +294,8 @@
 
         <python.version>3.7.5</python.version>
         <cpython-platform.version>${python.version}-${javacpp-presets.version}</cpython-platform.version>
+        <numpy.version>1.17.3</numpy.version>
```
Comment (saudet) [inline pom.xml:297]: Ah, you're actually adding that to this PR, that's why I missed it. Looks good

**Comment by AlexDBlack** [comment]
I can confirm this is running OK for me now on Windows, no more cache corruption issue.

**Comment by AlexDBlack** [comment]
@farizrahman4u are we good to go to merge this?

**Comment by farizrahman4u** [comment]
yes :shipit:

**Pull request merged**